File: Solutions/2487- Remove Nodes From Linked List/Solution.java
class Solution {
    public ListNode removeNodes(ListNode head) {
        Deque<Integer> stack = new ArrayDeque<>();
        var current = head;

        while (current != null) {
            while (!stack.isEmpty() && current.val > stack.peek()) {
                stack.pop();
            }
            stack.push(current.val);
            current = current.next;
        }
        ListNode answer = new ListNode();
        current = answer;
        while (!stack.isEmpty()) {
            current.next = new ListNode(stack.removeLast());
            current = current.next;
        }
        return answer.next;
    }

}

File: Solutions/2487- Remove Nodes From Linked List/Solution.py
class Solution:
    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        stack = []
        current = head

        while current is not None:
            while stack and current.val > stack[-1]:
                stack.pop()
            stack.append(current.val)
            current = current.next

        dummy = ListNode()
        current = dummy
        for val in stack:
            current.next = ListNode(val)
            current = current.next

        return dummy.next